src/prof_data.c

```diff
@@ -1444,9 +1441,21 @@ prof_tctx_destroy(tsd_t *tsd, prof_tctx_t *tctx) {
 	assert(tctx->cnts.accumobjs == 0);
 	assert(tctx->cnts.accumbytes == 0);
 
-	ckh_remove(tsd, &tdata->bt2tctx, &gctx->bt, NULL, NULL);
-	destroy_tdata = prof_tdata_should_destroy(tsd_tsdn(tsd), tdata, false);
-	malloc_mutex_unlock(tsd_tsdn(tsd), tdata->lock);
+	prof_gctx_t *gctx = tctx->gctx;
+
+	{
+		prof_tdata_t *tdata = tctx->tdata;
+		tctx->tdata = NULL;
```

Comment:
Question: what if tctx and tdata are both not destroyed? Should we keep tctx->tdata around so that we can try destroy again? I think the answer is no need, since we did the ckh_remove above?

Reply:
Right. The `ckh_remove()` decouples the `tctx` from the `tdata`. If the `tdata` ends up not being destroyed in this function, it could be destroyed later, but definitely not through this particular `tctx`. The `tctx->tdata = NULL` and `ckh_remove()` as a pair serve to completely detach the two.

Reply:
Thanks for the answers Yinan!